rlog/2024-04-20-rln-light-verifiers.mdx

```diff
@@ -0,0 +1,173 @@
+it costs approximately 2 Billion Gas.
+With Infura and Alchemy capping the gas limit to 350M and 550M gas respectively,
+it is not possible to compute the root of the tree in a single call.
+
+Acknowledging that [Polygon Miden](https://polygon.technology/blog/polygon-miden-state-model) and [Penumbra](https://penumbra.zone/blog/tiered-commitment-tree/) both make use of a tiered commitment tree,
+we propose a similar approach for RLN.
+
+A tiered commitment tree is a tree which is sharded into multiple smaller subtrees,
+each of which is a tree in itself.
+This allows scaling in terms of the number of leaves,
+as well as reducing state bloat,
+by just storing the root of a subtree when it is full,
+instead of all its leaves.
+
+Here, the main question arises:
+What is the maximum number of leaves in a subtree, with which the root can be computed in a single call?
+
+To compute the root of a Merkle tree with $2^{10}$ leaves, it costs approximately 217M gas.
+
+This is a feasible number for a single call,
+and hence we propose a tiered commitment tree with a maximum of $2^{10}$ leaves in a subtree,
+and the number of subtrees is $2^{10}$.
+Therefore, the maximum number of leaves in the tree is $2^{20}$,
+the same as the current implementation.
+
+![img](/img/light-rln-verifiers.png)
+
+### Insertion
+
+When a commitment is inserted into the tree,
+it is first inserted into the first subtree.
+When the first subtree is full,
+the next insertions go into the second subtree,
+and so on.
+
+### Syncing
+
+When syncing the tree,
+one only needs to fetch the roots of the subtrees,
+and the root of the tree can be computed in-memory, or on-chain.
+
+Which allows us to derive the following relation:
+
+$$
+number\_of\_rpc\_calls = number\_of\_filled\_subtrees + 1
+$$
+
+This is a significant improvement over the current implementation,
+which requires fetching all the memberships from the smart contract.
+
+### Gas costs
+
+The gas costs for inserting a commitment into the tree are the same as the current implementation,
+except it consists of an extra SSTORE operation to store the `shardIndex` of the commitment.
+
+### Events
+
+The events emitted by the contract are the same as the current implementation,
+appending the `shardIndex` of the commitment.
+
+### Proof of concept
+
+A proof of concept implementation of the tiered commitment tree is available [here](https://github.com/vacp2p/rln-contract/pull/37),
+and is deployed on Sepolia at [0xE7987c70B54Ff32f0D5CBbAA8c8Fc1cAf632b9A5](https://sepolia.etherscan.io/address/0xE7987c70B54Ff32f0D5CBbAA8c8Fc1cAf632b9A5).
+
+It is compatible with the current implementation of the RLN verifier.
+
+## Future work
+
+1. Optimize the gas costs of the tiered commitment tree.
+2. Explore using different number of leaves under a given node in the tree, currently set to 2.
+
+## Conclusion
+
+The tiered commitment tree is a promising approach to reduce the number of RPC calls required to sync the tree,
```

Comment:
Drop comma between 'sync the tree' and 'and to reduce'
Change 'and to reduce' to 'and reduce'